test/test_torch.py

```diff
@@ -5431,34 +5431,48 @@ def test(x: torch.Tensor, ia: torch.Tensor, ib: torch.Tensor) -> None:
 
         cpu = torch.device('cpu')
         for device in devices:
-            # Index cpu tensor with device tensor
             x = torch.randn(3, 4, 4, 4, 3)
-            ia = torch.tensor([0, 2, 1]).to(device)
-            ib = torch.tensor([0, 2, 1]).to(device)
-            test(x, ia, ib)
+            ia = torch.tensor([0, 2, 1])
+            ib = torch.tensor([0, 2, 1])
 
             # Index device tensor with cpu tensor
             x = x.to(device)
             ia = ia.to(cpu)
             ib = ib.to(cpu)
             test(x, ia, ib)
 
-            # Index cpu tensor with mixed cpu, device tensors
-            x = x.to(cpu)
+            # Index device tensor with mixed cpu, device tensors
+            x = x.to(device)
             ia = ia.to(cpu)
             ib = ib.to(device)
             test(x, ia, ib)
 
-            # Index device tensor with mixed cpu, device tensors
-            x = x.to(device)
+    @deviceCountAtLeast(1)
+    def test_advancedindex_mixed_devices_error(self, devices) -> None:
+        def test(x: torch.Tensor, ia: torch.Tensor, ib: torch.Tensor) -> None:
+            # test getitem
+            with self.assertRaisesRegex(RuntimeError, fr"indices should be either .* \({x.device}\)"):
+                value = x[:, ia, None, ib, 0]
+            with self.assertRaisesRegex(RuntimeError, fr"indices should be either .* \({x.device}\)"):
+                value = x[ib]
+
+        cpu = torch.device('cpu')
+        for device in devices:
+            # Index cpu tensor with device tensor
+            x = torch.randn(3, 4, 4, 4, 3)
+            ia = torch.tensor([0, 2, 1]).to(device)
+            ib = torch.tensor([0, 2, 1]).to(device)
+            test(x, ia, ib)
```

Comment:
XLA tests seem to be failing here, and I'm not sure why. Is there any possibility that after calling `tensor.to('xla')`, they end up on CPU?

Reply:
no, shouldn't be possible IMO

Reply:
I think so.
Another thing I think could be happening is that when `x[:, ia, None, ib, 0]` is executed, there's some (maybe fallback) dispatch step in XLA that: (i) moves those tensors to CPU; and (ii) dispatches to the CPU kernel (probably not the problem here, though).
I remember that, sometimes, XLA fallbacks to CPU, right? Is this something that could be happening? One easy fix to this is by making this test `@onlyCUDA`. What do you think?

Reply:
XLA doesn't share error checking logic with our kernels, so their index kernel probably needs an update to check that base + indices are all on the same device.

That logic is [here](https://github.com/pytorch/xla/blob/8ffef19735f8fa3278f4d42c8d64e86c31af49dd/torch_xla/csrc/aten_xla_type.cpp#L1600) - I can make a patch for it

Reply:
Thanks for looking it up, Brian.
But, on second thoughts, I think that will only be a problem after XLA's signature also changes to the **new** one. Here's what I think is happening:

- `at::index` tries to dispatch to the XLA kernel
- There's no XLA kernel that matches the **new** signature (with `IOptTensorRefList`)
- It gets dispatched to the CPU kernel via calling `xla_cpu_fallback`

Let me know what you think.

Reply:
Nope. Forget about my last reply.
You were totally right, Brian. That's what's wrong with this CI failure. Could you make a patch for it?
What I said is real only for #73351.

Reply:
Patch is here: https://github.com/pytorch/xla/pull/3414. I'll be on PTO for the rest of this week, but I can help land the patch when I get back (unless we need to land sooner, then maybe @JackCaoG can help land).